client/components/LearnerViewDirectiveSpec.js

```diff
@@ -0,0 +1,204 @@
+    // it.
+    // TODO (mengchaowang): Refactor learnerViewDirective controller to a
+    // separate controller instead of anonymous controller.
+    element = angular.element('<learner-view></learner-view>');
+    element = $compile(element)($scope);
+    $scope = element.isolateScope();
+    $rootScope.$digest();
+    $scope.$digest();
+
+    QuestionDataService = _QuestionDataService_;
+
+    SECONDS_TO_MILLISECONDS = _SECONDS_TO_MILLISECONDS_;
+    DEFAULT_AUTOSAVE_SECONDS = _DEFAULT_AUTOSAVE_SECONDS_;
+    AUTOSAVE_MILLISECONDS =
+      DEFAULT_AUTOSAVE_SECONDS * SECONDS_TO_MILLISECONDS;
+
+    localStorage.clear();
+
+
+  }));
+
+  describe("resetCode", function() {
+
+    it('should reset code to starter code', function() {
+      $scope.questionIds.forEach(function(questionId, index) {
+        var question = QuestionDataService.getQuestion(questionId);
+        var starterCode = question.getStarterCode(LANGUAGE);
+        $scope.currentQuestionIndex = index;
+        $scope.code = generateRandomChars(NUM_CHARS_CODE);
+        expect(angular.equals(starterCode, $scope.code)).toBe(false);
+        $scope.resetCode();
+        expect($scope.code).toEqual(starterCode);
+      });
+    });
+  });
+
+  describe("activateAutosaving", function() {
+    var $interval;
+    var $timeout;
+    var CodeStorageService;
+    beforeEach(inject(function(_$interval_, _$timeout_, _CodeStorageService_) {
+      $interval = _$interval_;
+      $timeout = _$timeout_;
+      CodeStorageService = _CodeStorageService_;
+      localStorage.clear();
+    }));
+
+    // Autosave text is using $timeout, thus,
+    // we need to flush both $interval and $timeout.
+    var flushIntervalAndTimeout = function(timeToFlush) {
+      $interval.flush(timeToFlush);
+      $timeout.flush(timeToFlush);
+    };
+
+    it("should only activate autosave once", function() {
+      for (var i = 0; i < $scope.questionIds.length; i++) {
+        if (i !== 0) {
+          // Default question index is 0, we don't want to
+          // navigate to question 0 again because navigateToQuestion
+          // will already have stored code for that question and fail the test.
+          $scope.navigateToQuestion(i);
+        }
+        expect(!$scope.autosaveOn).toBe(true);
+        var questionId = $scope.questionIds[i];
+        var question = QuestionDataService.getQuestion(questionId);
+        var starterCode = question.getStarterCode(LANGUAGE);
+        // There should be no code stored in localStorage
+        // before autosave is triggered.
+        expect(CodeStorageService.loadStoredCode(
+          questionId, LANGUAGE)).toEqual(null);
+        $scope.activateAutosaving();
+        checkActivateAutosaveOnceDetail(questionId, starterCode);
+        expect(CodeStorageService.loadStoredCode(
+          questionId, LANGUAGE)).toEqual(starterCode);
+      }
+    });
+
+    // This helper function is to test when autosave is triggered, what will
```

Comment:
Could you clean up a bit of the grammar / language, here? This comment is pretty verbose.